In [1]:
#IMPORTACIONES Y RUTAS
from pathlib import Path

import pandas as pd


# Permite ejecutar el notebook desde la carpeta notebooks
PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()

BREAST_CANCER_PATH = PROJECT_ROOT / "data" / "raw" / "breast_cancer.csv"
WINE_QUALITY_PATH = PROJECT_ROOT / "data" / "raw" / "wine_quality.csv"

if not BREAST_CANCER_PATH.exists():
    raise FileNotFoundError(f"No se encuentra: {BREAST_CANCER_PATH}")

if not WINE_QUALITY_PATH.exists():
    raise FileNotFoundError(f"No se encuentra: {WINE_QUALITY_PATH}")

In [2]:
#CARGAMOS LOS ARCHIVOS
breast_cancer_df = pd.read_csv(BREAST_CANCER_PATH)
wine_quality_df = pd.read_csv(WINE_QUALITY_PATH)

print("Breast Cancer:", breast_cancer_df.shape)
print("Wine Quality:", wine_quality_df.shape)

Breast Cancer: (569, 33)
Wine Quality: (1143, 13)


In [3]:
#VISUALIZAMOS LA ESTRUCTURA DE LOS DATOS
display(breast_cancer_df.head())
display(wine_quality_df.head())

print("Columnas de Breast Cancer:")
print(breast_cancer_df.columns.tolist())

print("\nColumnas de Wine Quality:")
print(wine_quality_df.columns.tolist())

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,NaN
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,NaN
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,NaN
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,NaN
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,NaN


,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,0
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,1
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,2
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,3
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,4


Columnas de Breast Cancer:
['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean', 'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean', 'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean', 'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se', 'fractal_dimension_se', 'radius_worst', 'texture_worst', 'perimeter_worst', 'area_worst', 'smoothness_worst', 'compactness_worst', 'concavity_worst', 'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32']

Columnas de Wine Quality:
['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality', 'Id']


In [4]:
#FUNCIÓN DE ANÁLISIS GENERAL
def summarize_dataset(
    dataframe: pd.DataFrame,
    name: str,
    target_column: str,
) -> None:
    """Muestra la información básica necesaria para validar un dataset."""

    print("=" * 70)
    print(name)
    print("=" * 70)

    print(f"Muestras: {dataframe.shape[0]}")
    print(f"Columnas: {dataframe.shape[1]}")
    print(f"Duplicados: {dataframe.duplicated().sum()}")
    print(f"Valores ausentes totales: {dataframe.isna().sum().sum()}")

    print("\nTipos de datos:")
    print(dataframe.dtypes.value_counts())

    print("\nValores ausentes por columna:")
    missing_values = dataframe.isna().sum()
    print(missing_values[missing_values > 0])

    print("\nDistribución de la variable objetivo:")
    print(dataframe[target_column].value_counts().sort_index())

    print("\nDistribución porcentual:")
    print(
        dataframe[target_column]
        .value_counts(normalize=True)
        .sort_index()
        .mul(100)
        .round(2)
    )

In [8]:
#ANALIZAMOD AMBOS DATASETS
summarize_dataset(
    dataframe=breast_cancer_df,
    name="Breast Cancer Wisconsin",
    target_column="diagnosis",
)

summarize_dataset(
    dataframe=wine_quality_df,
    name="Wine Quality",
    target_column="quality",
)

Breast Cancer Wisconsin
Muestras: 569
Columnas: 33
Duplicados: 0
Valores ausentes totales: 569

Tipos de datos:
float64    31
int64       1
str         1
Name: count, dtype: int64

Valores ausentes por columna:
Unnamed: 32    569
dtype: int64

Distribución de la variable objetivo:
diagnosis
B    357
M    212
Name: count, dtype: int64

Distribución porcentual:
diagnosis
B    62.74
M    37.26
Name: proportion, dtype: float64
Wine Quality
Muestras: 1143
Columnas: 13
Duplicados: 0
Valores ausentes totales: 0

Tipos de datos:
float64    11
int64       2
Name: count, dtype: int64

Valores ausentes por columna:
Series([], dtype: int64)

Distribución de la variable objetivo:
quality
3      6
4     33
5    483
6    462
7    143
8     16
Name: count, dtype: int64

Distribución porcentual:
quality
3     0.52
4     2.89
5    42.26
6    40.42
7    12.51
8     1.40
Name: proportion, dtype: float64


In [6]:
#VEMOS LAS COLUYMNAS QUE NO QUEREMOS

#Breast
breast_columns_to_remove = [
    column
    for column in ["id", "Unnamed: 32"]
    if column in breast_cancer_df.columns
]

print("Columnas que se eliminarán:", breast_columns_to_remove)

#Wine
wine_columns_to_remove = [
    column
    for column in ["Id", "id"]
    if column in wine_quality_df.columns
]

print("Columnas que se eliminarán:", wine_columns_to_remove)


Columnas que se eliminarán: ['id', 'Unnamed: 32']
Columnas que se eliminarán: ['Id']


In [7]:
#DISENO DE SELECCIÓN DE CATEGORÍAS PARA WINE
selected_wine_classes = [5, 6, 7]

wine_selected_df = wine_quality_df[
    wine_quality_df["quality"].isin(selected_wine_classes)
].copy()

print("Muestras originales:", len(wine_quality_df))
print("Muestras con clases 5, 6 y 7:", len(wine_selected_df))

print("\nDistribución:")
print(wine_selected_df["quality"].value_counts().sort_index())

print("\nDistribución porcentual:")
print(
    wine_selected_df["quality"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .round(2)
)

Muestras originales: 1143
Muestras con clases 5, 6 y 7: 1088

Distribución:
quality
5    483
6    462
7    143
Name: count, dtype: int64

Distribución porcentual:
quality
5    44.39
6    42.46
7    13.14
Name: proportion, dtype: float64


## Conclusiones de la exploración inicial

### Breast Cancer Wisconsin

- Dimensión original: ** 569 muestras y 33 columnas**.
- Variable objetivo: `diagnosis`.
- Clases observadas: `B` y `M`.
- Distribución de clases:
  - Benigno (`B`): 357 muestras.
  - Maligno (`M`): 212 muestras.
- Valores ausentes: 569.
- Registros duplicados: 0.
- Columnas que se eliminarán:
  - `id`
  - `Unnamed: 32`
- Número de características predictoras después de la limpieza: 30.

### Wine Quality

- Dimensión original: ** 1143 muestras y 13 columnas**.
- Variable objetivo: `quality`.
- Clases originales observadas: 6.
- Valores ausentes: 0.
- Registros duplicados: 0.
- Columna que se eliminará:
  - `Id`
- Clases seleccionadas para el experimento: 5, 6 y 7.
- Muestras después del filtrado: 1088.
- Distribución de las clases seleccionadas:
  - Calidad 5: 483 muestras.
  - Calidad 6: 462 muestras.
  - Calidad 7: 143 muestras.
- Número de características predictoras después de eliminar `Id`: 11.

### Decisiones preliminares

- Ambos datasets necesitan normalización.
- Las etiquetas deberán transformarse a valores numéricos.
- Se seleccionarán cuatro características para adaptar los datos al circuito de cuatro qubits.
- La necesidad de aplicar pesos de clase o equilibrado al dataset Wine Quality se decidirá a partir de la distribución observada.